File: edge_history.py
# -*- coding: utf-8 -*-
"""
FocusFlow Edge 浏览器历史记录模块

功能：
- 查询 Edge 浏览器总历史记录条数
- 查询指定日期的历史记录条数
- 按日期存储历史记录条数到本地数据库
- 支持查询近 N 天的趋势

Edge 历史记录存储在 SQLite 数据库中：
  %LOCALAPPDATA%\\Microsoft\\Edge\\User Data\\Default\\History
表 urls 记录所有访问的 URL，字段 last_visit_time 是 Chrome 时间戳
（从 1601-01-01 起的微秒数）
"""

import os
import sys
import sqlite3
import shutil
from datetime import datetime, date, timedelta, timezone
from typing import Optional, List, Tuple

from config import get_data_dir
from logger import get_logger

log = get_logger('edge_history')


def get_edge_history_path() -> str:
    """获取 Edge 历史记录数据库路径"""
    if sys.platform == 'win32':
        local_app = os.environ.get('LOCALAPPDATA', '')
        if local_app:
            return os.path.join(local_app, 'Microsoft', 'Edge', 'User Data', 'Default', 'History')
    return os.path.expanduser('~/.config/microsoft-edge/Default/History')


def _chrome_time_to_datetime(chrome_time: int) -> datetime:
    """Chrome 时间戳（微秒，从 1601-01-01 起）转 datetime"""
    epoch_start = datetime(1601, 1, 1, tzinfo=timezone.utc)
    return epoch_start + timedelta(microseconds=chrome_time)


def _datetime_to_chrome_time(dt: datetime) -> int:
    """datetime 转 Chrome 时间戳"""
    epoch_start = datetime(1601, 1, 1, tzinfo=timezone.utc)
    delta = dt - epoch_start
    return int(delta.total_seconds() * 1000000)


def _open_edge_history() -> Tuple[Optional[sqlite3.Connection], Optional[str]]:
    """打开 Edge 历史库，返回 (连接, 临时文件路径)

    策略：
    1. 优先只读直连（`file:...?mode=ro`）：Edge 未运行或允许读时，
       完全不复制大文件（Edge 历史库可达几十上百 MB），最快。
    2. 失败（Edge 运行中持有文件锁）则复制 主库+WAL+SHM 到临时文件后打开，
       保证包含 WAL 中未 checkpoint 的数据（旧版只复制主库会漏最近记录）。

    调用方负责：用完关闭连接，若临时文件非 None 则删除。
    """
    history_path = get_edge_history_path()
    if not os.path.exists(history_path):
        log.debug('Edge 历史记录文件不存在: %s', history_path)
        return None, None

    # 1) 只读直连
    try:
        conn = sqlite3.connect(f'file:{history_path}?mode=ro', uri=True)
        return conn, None
    except Exception as e:
        log.debug('Edge 历史库只读直连失败（可能被 Edge 锁定），改走复制: %s', e)

    # 2) 复制主库 + WAL + SHM（保证数据完整）
    temp_path = os.path.join(get_data_dir(), '_edge_history_temp.db')
    try:
        shutil.copy2(history_path, temp_path)
        for suffix in ('-wal', '-shm'):
            src = history_path + suffix
            if os.path.exists(src):
                try:
                    shutil.copy2(src, temp_path + suffix)
                except Exception:
                    pass
        conn = sqlite3.connect(temp_path)
        return conn, temp_path
    except Exception as e:
        log.error('复制 Edge 历史记录失败: %s', e)
        return None, None


def _close_edge_connection(conn, temp_path):
    """关闭连接并清理临时文件"""
    if conn is not None:
        try:
            conn.close()
        except Exception:
            pass
    if temp_path:
        for suffix in ('', '-wal', '-shm'):
            try:
                if os.path.exists(temp_path + suffix):
                    os.remove(temp_path + suffix)
            except Exception:
                pass


def query_edge_total_count() -> int:
    """查询 Edge 浏览器总历史记录条数

    Returns: 总历史记录条数
    """
    conn, temp_path = _open_edge_history()
    if conn is None:
        return 0
    try:
        cur = conn.execute('SELECT COUNT(*) FROM urls')
        return cur.fetchone()[0]
    except Exception as e:
        log.error('查询 Edge 总历史记录失败: %s', e)
        return 0
    finally:
        _close_edge_connection(conn, temp_path)


def query_edge_history_count(target_date: Optional[date] = None) -> int:
    """查询指定日期的 Edge 历史记录条数

    Args:
        target_date: 指定日期（None=今天）

    Returns: 该日期的历史记录条数
    """
    if target_date is None:
        target_date = date.today()

    conn, temp_path = _open_edge_history()
    if conn is None:
        return 0
    try:
        # 按本地时区计算日期范围
        local_tz = datetime.now().astimezone().tzinfo
        day_start = datetime.combine(target_date, datetime.min.time(), tzinfo=local_tz)
        day_end = day_start + timedelta(days=1)
        chrome_start = _datetime_to_chrome_time(day_start)
        chrome_end = _datetime_to_chrome_time(day_end)

        cur = conn.execute(
            'SELECT COUNT(*) FROM urls WHERE last_visit_time >= ? AND last_visit_time < ?',
            (chrome_start, chrome_end)
        )
        return cur.fetchone()[0]
    except Exception as e:
        log.error('查询 Edge 历史记录失败: %s', e)
        return 0
    finally:
        _close_edge_connection(conn, temp_path)


def save_edge_history_count(target_date: date, count: int):
    """保存指定日期的 Edge 历史记录条数到本地数据库"""
    db_path = os.path.join(get_data_dir(), 'focusflow_edge_history.db')
    try:
        conn = sqlite3.connect(db_path)
        conn.execute('''
            CREATE TABLE IF NOT EXISTS edge_history (
                date TEXT PRIMARY KEY,
                count INTEGER NOT NULL,
                updated_at INTEGER NOT NULL
            )
        ''')
        conn.execute(
            'INSERT OR REPLACE INTO edge_history (date, count, updated_at) VALUES (?, ?, ?)',
            (target_date.isoformat(), count, int(datetime.now().timestamp()))
        )
        conn.commit()
        conn.close()
        log.info('已保存 Edge 历史记录: %s = %d 条', target_date, count)
    except Exception as e:
        log.error('保存 Edge 历史记录失败: %s', e)


def get_edge_history_counts(days: int = 30) -> List[Tuple[str, int]]:
    """获取最近 N 天的 Edge 历史记录条数

    Returns: [(日期字符串, 条数), ...]
    """
    db_path = os.path.join(get_data_dir(), 'focusflow_edge_history.db')
    if not os.path.exists(db_path):
        return []

    try:
        conn = sqlite3.connect(db_path)
        conn.row_factory = sqlite3.Row
        now = datetime.now()
        start = now - timedelta(days=days - 1)
        cur = conn.execute(
            'SELECT date, count FROM edge_history WHERE date >= ? ORDER BY date',
            (start.date().isoformat(),)
        )
        result = [(row['date'], row['count']) for row in cur.fetchall()]
        conn.close()
        return result
    except Exception as e:
        log.error('查询 Edge 历史记录统计失败: %s', e)
        return []


def update_today_edge_history() -> Tuple[int, int]:
    """查询并保存今天的 Edge 历史记录

    Returns: (今日条数, 总条数)
    """
    today = date.today()
    today_count = query_edge_history_count(today)
    save_edge_history_count(today, today_count)
    total_count = query_edge_total_count()
    return today_count, total_count

File: crates/core/src/edge_history.rs
//! Edge 浏览器历史记录模块。
//!
//! 镜像 Python 版 `edge_history.py`：
//! - 直读 Edge History SQLite（%LOCALAPPDATA%\Microsoft\Edge\User Data\Default\History）
//! - Chrome 时间戳（1601-01-01 起微秒）转换
//! - 只读连接优先，被锁时复制 WAL 兜底
//! - 本地存储每日计数（focusflow_edge_history.db）供趋势图

use std::path::PathBuf;

use chrono::{Local, NaiveDate, TimeZone, Utc};
use rusqlite::Connection;

use crate::paths;

/// Edge History 数据库路径。
pub fn edge_history_path() -> PathBuf {
    if let Ok(local_app) = std::env::var("LOCALAPPDATA") {
        PathBuf::from(local_app)
            .join(r"Microsoft\Edge\User Data\Default\History")
    } else {
        PathBuf::from(r"C:\Users\Default\AppData\Local\Microsoft\Edge\User Data\Default\History")
    }
}

/// Chrome 时间戳（1601-01-01 起微秒）转 datetime。
#[allow(dead_code)]
fn chrome_to_datetime(chrome_time: i64) -> chrono::DateTime<Utc> {
    let epoch = Utc.with_ymd_and_hms(1601, 1, 1, 0, 0, 0).unwrap();
    epoch + chrono::Duration::microseconds(chrome_time)
}

/// datetime 转 Chrome 时间戳。
fn datetime_to_chrome(dt: &chrono::DateTime<Local>) -> i64 {
    let epoch = Utc.with_ymd_and_hms(1601, 1, 1, 0, 0, 0).unwrap();
    (dt.with_timezone(&Utc) - epoch).num_microseconds().unwrap_or(0)
}

/// 打开 Edge History（只读优先，锁定则复制）。
/// 复制前大小保护阈值：Edge 历史库超过该大小（字节）时不再复制
/// （复制几百 MB 会卡顿，直接返回错误提示）。
const EDGE_COPY_MAX_BYTES: u64 = 100 * 1024 * 1024; // 100MB

/// 直连 busy 等待上限：Edge 运行时会频繁短事务持排他锁，
/// 等待过短会让查询在锁窗口内失败并静默返回 0；过长则批量查询时逐个干等。
/// 被锁时 300ms 后自动转复制兜底（毫秒级），整体感知最快。
const EDGE_BUSY_TIMEOUT: std::time::Duration = std::time::Duration::from_millis(300);

/// 在 Edge History 上执行查询，返回 Option（None 表示读取失败/被锁）。
/// 策略：
/// 1) 只读直连（busy_timeout 1s）：Edge 未运行或锁间隙时最快。
/// 2) 查询失败（被锁）→ 复制主文件兜底重试；复制是整文件快照，
///    可能撞上 Edge 写事务产生撕裂快照，用少量重试覆盖。
/// 主文件在回滚日志模式下含全部已提交数据；WAL 模式下附带复制 -wal/-shm。
fn query_edge_count(query: impl Fn(&Connection) -> Option<i64>) -> Option<i64> {
    let path = edge_history_path();
    if !path.exists() {
        return None;
    }

    // 1) 只读直连：rusqlite 打开不一定失败（锁在首个查询才报 SQLITE_BUSY），
    //    必须实际验证查询可用才采用，否则 Edge 运行时直连会静默返回 0。
    if let Ok(conn) = Connection::open_with_flags(
        &path,
        rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY | rusqlite::OpenFlags::SQLITE_OPEN_NO_MUTEX,
    ) {
        let _ = conn.busy_timeout(EDGE_BUSY_TIMEOUT);
        if let Some(v) = query(&conn) {
            return Some(v);
        }
        drop(conn);
    }

    // 2) 复制兜底：先检查大小，超大库跳过复制避免卡顿
    let size = std::fs::metadata(&path).map(|m| m.len()).unwrap_or(0);
    if size > EDGE_COPY_MAX_BYTES {
        tracing::warn!(
            "Edge 历史库过大（{}MB），跳过复制（只读直连被锁定）",
            size / (1024 * 1024)
        );
        return None;
    }
    let temp = paths::data_dir().join("_edge_history_temp.db");
    for attempt in 0..3 {
        if std::fs::copy(&path, &temp).is_ok() {
            for suffix in ["-wal", "-shm"] {
                let src = format!("{}{}", path.display(), suffix);
                if std::path::Path::new(&src).exists() {
                    let _ = std::fs::copy(&src, format!("{}{}", temp.display(), suffix));
                }
            }
            if let Ok(conn) = Connection::open(&temp) {
                if let Some(v) = query(&conn) {
                    drop(conn);
                    let _ = std::fs::remove_file(&temp);
                    for suffix in ["-wal", "-shm"] {
                        let _ = std::fs::remove_file(format!("{}{}", temp.display(), suffix));
                    }
                    return Some(v);
                }
            }
        }
        tracing::debug!("Edge 历史库复制查询失败（第{}次），重试", attempt + 1);
        std::thread::sleep(std::time::Duration::from_millis(150));
    }
    let _ = std::fs::remove_file(&temp);
    for suffix in ["-wal", "-shm"] {
        let _ = std::fs::remove_file(format!("{}{}", temp.display(), suffix));
    }
    tracing::warn!("Edge 历史库无法读取（直连被锁且复制失败）");
    None
}

/// 查询 Edge 总历史记录数（失败/被锁返回 None）。
pub fn query_edge_total_count() -> Option<i64> {
    query_edge_count(|conn| {
        conn.query_row("SELECT COUNT(*) FROM urls", [], |r| r.get::<_, i64>(0))
            .ok()
    })
}

/// 查询指定日期的 Edge 历史记录数（失败/被锁返回 None）。
pub fn query_edge_history_count(target_date: NaiveDate) -> Option<i64> {
    let day_start = Local
        .from_local_datetime(&target_date.and_hms_opt(0, 0, 0).unwrap())
        .single()
        .unwrap();
    let day_end = day_start + chrono::Duration::days(1);
    let chrome_start = datetime_to_chrome(&day_start);
    let chrome_end = datetime_to_chrome(&day_end);
    query_edge_count(|conn| {
        conn.query_row(
            "SELECT COUNT(*) FROM urls WHERE last_visit_time >= ?1 AND last_visit_time < ?2",
            rusqlite::params![chrome_start, chrome_end],
            |r| r.get::<_, i64>(0),
        )
        .ok()
    })
}

fn edge_db_path() -> PathBuf {
    paths::data_dir().join("focusflow_edge_history.db")
}

/// 保存指定日期的计数到本地。
pub fn save_edge_history_count(target_date: NaiveDate, count: i64) {
    let conn = match Connection::open(edge_db_path()) {
        Ok(c) => c,
        Err(_) => return,
    };
    let _ = conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS edge_history (
            date TEXT PRIMARY KEY,
            count INTEGER NOT NULL,
            updated_at INTEGER NOT NULL
        );",
    );
    let _ = conn.execute(
        "INSERT OR REPLACE INTO edge_history (date, count, updated_at) VALUES (?1, ?2, ?3)",
        rusqlite::params![
            target_date.format("%Y-%m-%d").to_string(),
            count,
            Utc::now().timestamp()
        ],
    );
}

/// 获取近 N 天 Edge 历史计数。
pub fn get_edge_history_counts(days: i64) -> Vec<(String, i64)> {
    let path = edge_db_path();
    if !path.exists() {
        return Vec::new();
    }
    let conn = match Connection::open(&path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };
    let start = (Local::now().date_naive() - chrono::Days::new((days - 1).max(0) as u64))
        .format("%Y-%m-%d")
        .to_string();
    let result = conn
        .prepare("SELECT date, count FROM edge_history WHERE date >= ?1 ORDER BY date")
        .and_then(|mut stmt| {
            stmt.query_map([&start], |r| Ok((r.get::<_, String>(0)?, r.get::<_, i64>(1)?)))
                .map(|it| it.flatten().collect())
        });
    result.unwrap_or_default()
}

/// 更新今天并返回 (是否成功, 今日数, 总数)。
/// 任一步失败（Edge 库被锁/不可读）返回 (false, 0, 0)，调用方据此提示用户，
/// 避免把失败静默当成"0 条记录"。成功后后台补齐近 30 天缺失的历史计数。
pub fn update_today_edge_history() -> (bool, i64, i64) {
    let today = Local::now().date_naive();
    match (query_edge_history_count(today), query_edge_total_count()) {
        (Some(today_count), Some(total)) => {
            save_edge_history_count(today, today_count);
            save_edge_history_meta("today", today_count);
            save_edge_history_meta("total", total);
            // 后台补齐近 30 天缺失日期（不阻塞刷新返回）
            std::thread::Builder::new()
                .name("edge-backfill".into())
                .spawn(backfill_edge_history(30))
                .ok();
            (true, today_count, total)
        }
        _ => (false, 0, 0),
    }
}

/// 补齐近 N 天缺失的 Edge 历史计数（趋势表）。
/// 只查询本地库中还没有记录的天，避免每次全量重查。
fn backfill_edge_history(days: i64) -> impl FnOnce() + Send + 'static {
    move || {
        let today = Local::now().date_naive();
        let start = today - chrono::Days::new((days - 1).max(0) as u64);
        let existing: std::collections::HashSet<String> = Connection::open(edge_db_path())
            .ok()
            .and_then(|conn| {
                conn.prepare("SELECT date FROM edge_history WHERE date >= ?1")
                    .ok()
                    .and_then(|mut stmt| {
                        stmt.query_map([start.format("%Y-%m-%d").to_string()], |r| {
                            r.get::<_, String>(0)
                        })
                        .ok()
                        .map(|it| it.flatten().collect())
                    })
            })
            .unwrap_or_default();
        let mut filled = 0;
        let mut day = start;
        while day <= today {
            let key = day.format("%Y-%m-%d").to_string();
            if !existing.contains(&key) {
                if let Some(c) = query_edge_history_count(day) {
                    save_edge_history_count(day, c);
                    filled += 1;
                }
            }
            day = day + chrono::Days::new(1);
        }
        if filled > 0 {
            tracing::info!("Edge 历史已补齐 {} 天缺失记录", filled);
        }
    }
}

/// 保存上次刷新的数值（meta 表），插件重启后恢复显示，避免出现误导性的 "—" / 0。
fn save_edge_history_meta(key: &str, value: i64) {
    let conn = match Connection::open(edge_db_path()) {
        Ok(c) => c,
        Err(_) => return,
    };
    let _ = conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT NOT NULL);",
    );
    let _ = conn.execute(
        "INSERT OR REPLACE INTO meta (key, value) VALUES (?1, ?2)",
        rusqlite::params![key, value.to_string()],
    );
}

/// 读取上次保存的今日计数（本地缓存，未刷新过返回 None）。
pub fn get_edge_history_saved_today() -> Option<i64> {
    let conn = Connection::open(edge_db_path()).ok()?;
    conn.query_row(
        "SELECT value FROM meta WHERE key='today'",
        [],
        |r| r.get::<_, String>(0),
    )
    .ok()
    .and_then(|v| v.parse().ok())
}

/// 读取上次保存的总记录数（本地缓存，未刷新过返回 None）。
pub fn get_edge_history_saved_total() -> Option<i64> {
    let conn = Connection::open(edge_db_path()).ok()?;
    conn.query_row(
        "SELECT value FROM meta WHERE key='total'",
        [],
        |r| r.get::<_, String>(0),
    )
    .ok()
    .and_then(|v| v.parse().ok())
}

/// 本地趋势（近 N 天），供插件展示。
pub fn trend_counts(days: i64) -> Vec<(String, i64)> {
    get_edge_history_counts(days)
}
